### PR TITLE
Lock used UTXOs

### LINKED ISSUE #737: Got `bad-txns-inputs-missingorspent` when broadcasting lock transaction
This was a preview maker CFD, here are the logs:

```
2021-11-29 11:25:09  INFO Taking current order: Order { id: OrderId(Hyphenated(65f6b024-a631-4676-b8c7-136b5587337e)), trading_pair: BtcUsd, position: Short, price: Price(57516.6525), min_quantity: Usd(10), max_quantity: Usd(1000), leverage: Leverage(2), liquidation_price: Price(38344.4350), creation_timestamp: Timestamp(1638144405), settlement_interval: Duration { seconds: 86400, nanoseconds: 0 }, origin: Theirs, oracle_event_id: BitMexPriceEventId { timestamp: OffsetDateTime { utc_datetime: PrimitiveDateTime { date: Date { year: 2021, ordinal: 334 }, time: Time { hour: 1, minute: 0, second: 0, nanosecond: 0 } }, offset: UtcOffset { hours: 0, minutes: 0, seconds: 0 } }, digits: 20 } }
2021-11-29 11:25:11  INFO Order got accepted order_id=65f6b024-a631-4676-b8c7-136b5587337e
2021-11-29 11:25:12  INFO Exchanged setup parameters
2021-11-29 11:25:17  INFO Created CFD transactions
2021-11-29 11:25:18  INFO POST /api/calculate/margin application/json:    
2021-11-29 11:25:20 DEBUG Got notification for new block block_height=2105659
2021-11-29 11:25:25  INFO GET /alive:    
2021-11-29 11:25:29  INFO Exchanged CFD transactions
2021-11-29 11:25:35  INFO GET /alive:    
2021-11-29 11:25:40  INFO GET /alive:    
2021-11-29 11:25:50  INFO GET /alive:    
2021-11-29 11:25:51  INFO Verified all signatures
2021-11-29 11:25:51  INFO Exchanged signed lock transaction
2021-11-29 11:26:09  INFO Setup complete, publishing on chain now
2021-11-29 11:26:10 ERROR Message handler failed: Broadcasting transaction failed. Txid: 83fed9c0677fb30324494598bd5bf6e337f6a737d217893984bb6a4a5ff1ffd3. Raw transaction: 020000000001028d96dd5c90f512f36dc0e00da71b0f31a4ffbffd627b2ed5a17c34d75ef561d90000000000ffffffff92191b86ad41d9c25dafa276331debcd9bf3782f6b2cc6ac9a009c27669fe33f0100000000ffffffff03df65000000000000220020e00980a75447cfeb32059a78c7e5457fae209a2b88bc1055f5dae52c2d6c3491d151020000000000160014f32ec9844f11a0c09e491d8073a35f462234f8e955270100000000001600147ff6eac06f3db0e0db09d6458f88fba22dc6cd9a0247304402200610bc3a508f0e79b69725ce33c8508d750ad7b7a13ed4c1fcf2add622efe5d2022054c974f56d77beae0258305938809a90dbf3fcd97152348df99217e6dfe886820121027a317b986bceb60b5a65a71e11294ab96cdf10f0ae73998d7962b2370838be820247304402205fadab625736750a0da5174efe61866e3886b8c5ae6bbdb119810c55a2035c6b02201cee69e7c9470845c3c3da8c23471174cd1f69035a77b8d3ea9e483d653d4758012102be4bfaa2abac6a187c6d5432da5720a447b315928d3b56c8a9c335db9b9f7ee400000000: Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-25,\"message\":\"bad-txns-inputs-missingorspent\"}")))
```

### PR DESCRIPTION
**Pull request by thomaseizinger**
Fixes #737.

### CODE REVIEW

File: daemon/src/wallet.rs [hunk 1]
```diff
@@ -273,9 +273,55 @@ impl From<RpcErrorCode> for i64 {
...
+        &mut self,
+        amount: Amount,
+        used_utxos: &mut HashSet<OutPoint>,
+    ) -> Result<PartiallySignedTransaction> {
+        let mut builder = self.build_tx();
+
+        builder
+            .ordering(TxOrdering::Bip69Lexicographic) // TODO: I think this is pointless but we did this in maia.
+            .fee_rate(FeeRate::from_sat_per_vb(1.0))
+            .unspendable(used_utxos.iter().copied().collect())
```
Comment (bonomat) [inline daemon/src/wallet.rs:302]: Nice 👍

**Review by bonomat — APPROVED** [review]
I wonder if we should implement an automated cleanup though. At least for the maker this in-memory collection can grow. Compared to other things we store in memory this might be negligible though :D

**Comment by thomaseizinger** [comment]
> At least for the maker this in-memory collection can grow.

Yep, that is true!

How to do it reliably is hard though so I would like to postpone that. It is 33 bytes per UTXO. We can store around 30_000 UTXOs in memory and will still only leak 1 MB of RAM.

**Comment by bonomat** [comment]
> How to do it reliably is hard though so I would like to postpone that. It is 33 bytes per UTXO. We can store around 30_000 UTXOs in memory and will still only leak 1 MB of RAM.

sure, let's solve this once we run into this problem.

**Review by da-kami — APPROVED** [review]

File: daemon/src/wallet.rs [hunk 2]
```diff
@@ -285,4 +331,43 @@ mod tests {
...
+                .global
+                .unsigned_tx
+                .input
+                .iter()
+                .map(|i| i.previous_output),
+        );
+
+        // 2 TX a 2500 sats with UTXOs worth 1000s = 6 inputs
+        // If there are 6 UTXOs in the HashSet, we know that they are all different (HashSets don't
+        // allow duplicates!)
```
Comment (klochowicz) [inline daemon/src/wallet.rs:367]: great comment documenting the magic number 👍

**Review by klochowicz — APPROVED** [review]

**Pull request merged**